chore: linting setup

filoozom opened the pull request:
Tasks:
- [x] Lint all the code
- [x] Fix TypeScript errors
- [x] Fix TypeScript warnings (fixed most, still a few to go)

This is a pretty bad PR to review, I now realize I should have done the TypeScript things separately from the linting.

sripwoud commented:
Thanks for the PR!
Could you use https://github.com/zkitter/configs to stay consistent with the other zkitter repos?
Especially prettify the code rather with prettier.

> I should have done the TypeScript things separately from the linting

I think that's fine to include all those static analysis tests together.

sripwoud commented on .editorconfig:
```diff
@@ -0,0 +1,7 @@
+[*]
```
Interesting. I did not know https://editorconfig.org/
It's nice to have something that doesn't require a additional node package.
But we can't assume the dev is using a supported IDE (maybe some people use nano🤣?).
More importantly if you rely only on the IDE + `.editorconfig`, you won't be able to perform a formatting check in a CI workflow.
So please remove this file and rely only on `prettier` (`@zkitter/prettier-config`) instead

filoozom replied:
It's also super useful to set default line endings for example, and you can check that with CI afterwards if you wish to, but sure, I'll remove it 🙂

sripwoud commented on .eslintrc.json:
```diff
@@ -0,0 +1,27 @@
+{
+	"extends": [
+		"prettier",
+		"plugin:prettier/recommended",
```
Please do not use this

sripwoud commented on package.json:
```diff
@@ -1,115 +1,117 @@
+		"copy-webpack-plugin": "^11.0.0",
+		"eslint": "^8.35.0",
+		"eslint-config-prettier": "^8.5.0",
+		"eslint-plugin-prettier": "^4.2.1",
```
Related to previous [comment](https://github.com/zkitter/zkitter-js/pull/1/files#r1121978103).
I am not a fan of turning prettier rules into eslint rules, which is what `eslint-plugin-prettier` and `plugin:prettier/recommended` do. I prefer to just turn some linting rules off to avoid clashing with eslint.
Which is what prettier recommends in their [docs](https://prettier.io/docs/en/integrating-with-linters.html#notes).

filoozom replied:
Alright, sorry, wasn't aware of the configs repo!

I installed your plugins, but I can't get `eslint` to output something that makes sense. I get:
```
✖ 418 problems (284 errors, 134 warnings)
```

I'll dig into it a bit more and if I can't find why I'll push a wip commit so you can check it out.

filoozom commented:
Hi,

I think I'm done now, sorry for the very messy git history...
I rebased on latest `main` as well.

There's still a few warnings, mostly:
- `@typescript-eslint/no-non-null-assertion`: They should probably be checked and `throw` errors if variables are empty? Or enforce stricter types?
- `@typescript-eslint/no-explicit-any`: I fixed most of these, the remaining ones are some where I don't fully understand the code I think.
- `no-console`: I'll leave those up to you. I just disabled that rule in the `cli` directory.

Also one error:
- `no-async-promise-executor`: It's in `queryHistory`, but I didn't want to change actual code in this PR, only types.

I'd appreciate if this could be reviewed before adding other commits, because the rebase was a huge pain 😝

sripwoud commented:
@filoozom 

> The rebase was a huge pain.

I can imagine. I have to ask you to rebase one last time still.
I've just created the GH action that checks the linting etc https://github.com/zkitter/zkitter-js/pull/6
After rebasing we can then check the GH action results against your branch.

> There's still a few warnings, mostly:
Warnings are fine. Addressing them is not a must imo.

> Also one error:
> no-async-promise-executor: It's in queryHistory, but I didn't want to change actual code in this PR, only types.

What about eslint-lint ignoring it inline?
Or if we notice this linting error keeps coming, we can decide to override it in `.eslintrc.yaml`

filoozom commented:
> What about eslint-lint ignoring it inline?

I can, but I think it might be better to first merge this so we don't forget about the issue, and then fix it afterwards? I could also just fix it here, but ideally this should only do the formatting part.

EDIT: Actually I just fixed the `new Promise` linting error as it's straightforward.
EDIT 2: Disabled linting of `.js` files as that doesn't work with the current config.

sripwoud commented:
One last rebase please😁
I did some work on `main` to fix:
- formatting https://github.com/zkitter/zkitter-js/actions/runs/4348757668/jobs/7602900856#step:8:67
- missing used deps (the [failing depcheck](https://github.com/zkitter/zkitter-js/actions/runs/4348757668/jobs/7602900856#step:8:64) and all the [related tsc errors](https://github.com/zkitter/zkitter-js/actions/runs/4348757668/jobs/7602900856#step:8:78))
- errors about [not recognized use of accessors](https://github.com/zkitter/zkitter-js/actions/runs/4348757668/jobs/7602900856#step:8:69) (this is something I broke when splitting the tsconfig files, see https://github.com/zkitter/zkitter-js/commit/63d236fbdc20a6f68e28969608b8ed746f3e3d4f)

Then we should be good.
(It looks like there will still be some tsc errors but let's address them separately, this PR is indeed big enough)
[Example](https://github.com/zkitter/zkitter-js/actions/runs/4348757668/jobs/7602900856#step:8:87)
